sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -75,18 +76,35 @@ abstract class PlanExpression[T <: QueryPlan[_]] extends Expression {
  */
 abstract class SubqueryExpression(
     plan: LogicalPlan,
-    outerAttrs: Seq[Expression],
+    outerAttrs: Seq[(Expression, Boolean)],
```

Comment:
Let's say the plan is
```
OuterPlan1
  OuterPlan2
    SubqueryExpression(attr1, attr2)
```

If `attr1` is from `OuterPlan1`, then it should be `OuterReference`. If `attr2` is from `OuterPlan2`, it should be `AttributeReference`. The rationale is: `SubqueryExpression` lives in the `OuterPlan2`, so it can directly use `AttributeReference` to reference columns (but not in `SubqueryExpression#plan`).

Seems we don't need the boolean flag?

Reply:
The decorrelation framework needs to distinguish between normal outerAttrs and outerScopeAttrs because we have some extra operations for outerScopeAttrs.

For this case, both attr2 and att1 should be OuterReference as attr2 can't be resolved in the SubqueryExpression.plan.

Reply:
If you mean that attr2 can be AttributeReference and attr1 can be OuterReference(attr1), it's not that clear comparing to boolean flags. And later in the optimizer, we have stripOuterReference process for the SubqueryExpression and also we have checks to see if a expression/plan have OuterReference to decide whether the subquery plan is correlated but not processed or not. Having additional OuterWrapper in the SubqueryExpression is very misleading.